OdbBackend: initial groundwork

ddevault opened the pull request:
The rough plan is to make the `OdbBackend` type a fairly straightforward mirror of the `odb_backend` in libgit2. I'll then be adding `LooseOdbBackend` and `PackOdbBackend`, and `MempackOdbBackend`, which will inherit from `OdbBackend` and set the underlying `odb_backend` pointer, which should Just Werk based on the approach I have set out so far.

The more interesting case is that of user-created `OdbBackend` implementations. I'm not really sure what to do for this yet. The first inch of this work is done in the form of the read call raising `NotImplementedException` if the C implementation is NULL, then through some kind of magic I assume I'll be able to set these to function pointers for some kind of proxy into Python code with some kind of magic base class. Or something. ¯\\\_(ツ)\_/¯

ddevault commented on src/odb_backend.c:
```diff
@@ -0,0 +1,142 @@
+    "Read raw object data from this odb backend.\n");
+
+PyObject *
+OdbBackend_read(OdbBackend *self, PyObject *py_hex)
```
Note, in keeping with the libgit2 API design, I have no intention of refactoring `Odb_read` to go through this code for the sake of DRY.

ddevault commented on src/odb_backend.c:
```diff
@@ -0,0 +1,142 @@
+
+    /* XXX: This assumes the default libgit2 allocator is in use and will
+     * probably segfault and/or destroy the universe otherwise */
+    free(data);
```
:hankey:

jdavid commented on src/odb.c:
```diff
@@ -234,6 +237,80 @@ Odb_write(Odb *self, PyObject *args)
+    for (i = 0; i < nbackends; ++i) {
+        err = git_odb_get_backend(&backend, self->odb, i);
+        if (err != 0)
+            return Error_set(err);
```
If there's an error you've to `Py_DECREF(accum)` to avoid leaking.

jdavid commented on src/odb.c:
```diff
@@ -234,6 +237,80 @@ Odb_write(Odb *self, PyObject *args)
+    int err;
+    size_t i, nbackends;
+    git_odb_backend *backend;
+    PyObject *accum = PyList_New(0);
```
Check for errors, if `accum` is `NULL`

jdavid commented:
looks good
just made a couple of small comments regarding error checking and handling

ddevault commented:
I think the basics are in good shape here. I'm going to implement the rest of the backend interface next and then figure out how to make new backends in Python in a future pull request.

jdavid commented on src/odb_backend.c:
```diff
@@ -0,0 +1,348 @@
+
+    err = PyList_Append((PyObject*)accum, py_oid);
+    Py_DECREF(py_oid);
+    return err;
```
Return `GIT_EUSER` if there's an error.
`PyList_Append` will return -1 on error, which would be interpreted as `GIT_ERROR` in `OdbBackend_as_iter`, and the exception set by `PyList_Append` would be overwritten.

More information in https://github.com/libgit2/libgit2/issues/4783

`Odb_build_as_iter` has the same problem, please fix it there as well.

jdavid commented on src/odb_backend.c:
```diff
@@ -0,0 +1,348 @@
+    if (err == GIT_EUSER)
+        return NULL;
+    if (err < 0)
+        return Error_set(err);
```
On these 2 errors you've to `Py_DECREF(accum);`.
Same problem in `Odb_as_iter`

jdavid commented on src/odb_backend.c:
```diff
@@ -0,0 +1,348 @@
+OdbBackend_as_iter(OdbBackend *self)
+{
+    int err;
+    PyObject *accum = PyList_New(0);
```
Check for errors returned by `PyList_New`
Same problem in Odb_as_iter

jdavid commented on src/odb_backend.c:
```diff
@@ -0,0 +1,348 @@
+        "(ns#)",
+    #else
+        "(ny#)",
+    #endif
```
I've removed support for Python 2 in master, so you can update this.

jdavid commented on src/odb_backend.c:
```diff
@@ -0,0 +1,348 @@
+        "Oii|II";
+    #else
+        "Oip|II";
+    #endif
```
I've removed support for Python 2 in master, so you can update this.

jdavid commented:
I've fixed the build errors in AppVeyor, now there're some test failures related to this work.

ddevault commented:
I've addressed the feedback and I think that this is mergable, except for the build failure. I looked into it and I don't really understand why it's occuring. By my reckoning, this should affect a lot of tests. Can you take a look and tell me what you think?

jdavid commented:
We've seen these errors before.

Looking for a little bit I've found commit 00dd78bf1b3 which fixed 7 errors (6 of them were PermissionErrors): https://ci.appveyor.com/project/techtonik/pygit2/build/1.0.12/tests

Commit af983b9 is likely related as well.

ddevault commented:
Thanks!
